File: examples/python/02_robot_model_setting.py
# Robot Model Setting Example
#
# This example includes the following features:
# 1. Loading the current robot model (in URDF format) from the robot.
# 2. Saving a custom robot model to the robot with a specified name.
# 3. Assigning the robot model name for the robot to use (applied after reboot).
#
# Note:
# If the modified URDF causes any issue, users can recover by downloading the
# proper official URDF for their robot model from the rby1-sdk repository below,
# parsing it in the same way as this example, and importing it again with a new model name:
# https://github.com/RainbowRobotics/rby1-sdk/tree/main/models
#
# If an issue arises while modifying the urdf again after the above step,
# You can skip the previous step by running only the robot.set_parameter() command.
#
# Copyright (c) 2025 Rainbow Robotics. All rights reserved.
#
# DISCLAIMER:
# This is a sample code provided for educational and reference purposes only.
# Rainbow Robotics shall not be held liable for any damages or malfunctions resulting from
# the use or misuse of this demo code. Please use with caution and at your own discretion.

import rby1_sdk
import argparse
import xml.etree.ElementTree as ET


def main(address, model):
    robot = rby1_sdk.create_robot(address, model)
    robot.connect()
    if not robot.is_connected():
        print("Failed to connect robot")
        exit(1)

    # 1. Load current robot model from robot
    urdf_str = robot.get_robot_model()

    print("Current robot model loaded.")

    # 2. Parse URDF
    model_tree = ET.ElementTree(ET.fromstring(urdf_str))
    model_root = model_tree.getroot()

    # 3. Modify ee_right mass
    target_link_name = "ee_right"
    mass_delta = 0.5  # increase by 0.1 kg

    modified = False

    for link in model_root.findall("link"):
        if link.get("name") == target_link_name:
            inertial = link.find("inertial")
            if inertial is None:
                print(f"[ERROR] link '{target_link_name}' has no <inertial> tag.")
                exit(1)

            mass_tag = inertial.find("mass")
            if mass_tag is None:
                print(f"[ERROR] link '{target_link_name}' has no <mass> tag.")
                exit(1)

            old_mass = float(mass_tag.get("value"))
            new_mass = old_mass + mass_delta
            mass_tag.set("value", f"{new_mass:.8f}")

            print(f"Modified '{target_link_name}' mass: {old_mass:.8f} -> {new_mass:.8f}")
            modified = True
            break

    if not modified:
        print(f"[ERROR] link '{target_link_name}' not found in URDF.")
        exit(1)

    # 4. Upload modified model with a new name
    new_model_name = "temp_ee_right_mass_up"
    modified_urdf = ET.tostring(model_root, encoding="unicode")

    ok = robot.import_robot_model(new_model_name, modified_urdf)
    print(f"Import robot model: {ok}")

    if not ok:
        print("[ERROR] Failed to import modified robot model.")
        exit(1)

    # 5. Set robot model name
    robot.set_parameter("model_name", f'"{new_model_name}"')
    print(f"Set model_name to '{new_model_name}'")
    print("The modified robot model will be applied after reboot.")


if __name__ == "__main__":
    parser = argparse.ArgumentParser(description="Modify ee_right mass example")
    parser.add_argument("--address", type=str, required=True, help="Robot address")
    parser.add_argument("--model", type=str, default="a", help="Robot Model Name (default: 'a')")
    args = parser.parse_args()

    main(address=args.address, model=args.model)

File: examples/cpp/02_robot_model_setting.cpp
// Robot Model Setting Example

// This example includes the following features:
// 1. Loading the current robot model (in URDF format) from the robot.
// 2. Saving a custom robot model to the robot with a specified name.
// 3. Assigning the robot model name for the robot to use (applied after reboot).

// Note:
// If the modified URDF causes any issue, users can recover by downloading the
// proper official URDF for their robot model from the rby1-sdk repository below,
// parsing it in the same way as this example, and importing it again with a new model name:
// https://github.com/RainbowRobotics/rby1-sdk/tree/main/models

// If an issue arises while modifying the urdf again after the above step,
// You can skip the previous step by running only the robot.set_parameter() command.
//
// Usage example:
//   ./example_02_robot_model --address 192.168.30.1:50051 --model a
//
// Copyright (c) 2025 Rainbow Robotics. All rights reserved.
//
// DISCLAIMER:
// This is a sample code provided for educational and reference purposes only.
// Rainbow Robotics shall not be held liable for any damages or malfunctions resulting from
// the use or misuse of this demo code. Please use with caution and at your own discretion.

#include <algorithm>
#include <cctype>
#include <iomanip>
#include <iostream>
#include <sstream>
#include <string>

#include "tinyxml2.h"

#include "rby1-sdk/model.h"
#include "rby1-sdk/robot.h"

using namespace rb;

namespace {

std::string ToLower(std::string v) {
  std::transform(v.begin(), v.end(), v.begin(), [](unsigned char c) { return static_cast<char>(std::tolower(c)); });
  return v;
}

void PrintUsage(const char* prog) {
  std::cerr << "Usage: " << prog << " --address <server address> [--model a|m|ub]" << std::endl;
  std::cerr << "   or: " << prog << " <server address> [model]" << std::endl;
}

template <typename ModelT>
int Run(const std::string& address) {
  auto robot = Robot<ModelT>::Create(address);
  robot->Connect();
  if (!robot->IsConnected()) {
    std::cerr << "Failed to connect robot" << std::endl;
    return 1;
  }

  // 1. Load current robot model from robot
  const std::string urdf_str = robot->GetRobotModel();
  std::cout << "Current robot model loaded." << std::endl;

  // 2. Parse URDF
  tinyxml2::XMLDocument doc;
  if (doc.Parse(urdf_str.c_str()) != tinyxml2::XML_SUCCESS) {
    std::cerr << "Failed to parse robot model XML" << std::endl;
    return 1;
  }

  auto* root = doc.RootElement();
  if (!root) {
    std::cerr << "Invalid robot model XML (no root element)" << std::endl;
    return 1;
  }

  // 3. Modify ee_right mass
  const std::string target_link_name = "ee_right";
  constexpr double mass_delta = 0.5;
  bool modified = false;

  for (auto* link = root->FirstChildElement("link"); link; link = link->NextSiblingElement("link")) {
    const char* name = link->Attribute("name");
    if (name && std::string(name) == target_link_name) {
      auto* inertial = link->FirstChildElement("inertial");
      if (!inertial) {
        std::cerr << "[ERROR] link '" << target_link_name << "' has no <inertial> tag." << std::endl;
        return 1;
      }

      auto* mass_tag = inertial->FirstChildElement("mass");
      if (!mass_tag) {
        std::cerr << "[ERROR] link '" << target_link_name << "' has no <mass> tag." << std::endl;
        return 1;
      }

      double old_mass = 0.0;
      if (mass_tag->QueryDoubleAttribute("value", &old_mass) != tinyxml2::XML_SUCCESS) {
        std::cerr << "[ERROR] link '" << target_link_name << "' has invalid <mass value>." << std::endl;
        return 1;
      }

      const double new_mass = old_mass + mass_delta;
      std::ostringstream mass_ss;
      mass_ss << std::fixed << std::setprecision(8) << new_mass;
      mass_tag->SetAttribute("value", mass_ss.str().c_str());

      std::cout << std::fixed << std::setprecision(8) << "Modified '" << target_link_name
                << "' mass: " << old_mass << " -> " << new_mass << std::endl;
      modified = true;
      break;
    }
  }

  if (!modified) {
    std::cerr << "[ERROR] link '" << target_link_name << "' not found in URDF." << std::endl;
    return 1;
  }

  // 4. Upload modified model with a new name
  const std::string new_model_name = "temp_ee_right_mass_up";
  tinyxml2::XMLPrinter printer;
  doc.Print(&printer);
  const bool ok = robot->ImportRobotModel(new_model_name, printer.CStr());
  std::cout << "Import robot model: " << std::boolalpha << ok << std::endl;
  if (!ok) {
    std::cerr << "[ERROR] Failed to import modified robot model." << std::endl;
    return 1;
  }

  // 5. Set robot model name (applied after reboot)
  if (!robot->SetParameter("model_name", "\"" + new_model_name + "\"")) {
    std::cerr << "[ERROR] Failed to set model_name parameter." << std::endl;
    return 1;
  }
  std::cout << "Set model_name to '" << new_model_name << "'" << std::endl;
  std::cout << "The modified robot model will be applied after reboot." << std::endl;

  return 0;
}

}  // namespace

int main(int argc, char** argv) {
  std::string address;
  std::string model = "a";

  for (int i = 1; i < argc; ++i) {
    std::string arg = argv[i];
    if (arg == "--address" && i + 1 < argc) {
      address = argv[++i];
    } else if (arg == "--model" && i + 1 < argc) {
      model = argv[++i];
    } else if (arg.rfind("--", 0) == 0) {
      PrintUsage(argv[0]);
      return 1;
    } else if (address.empty()) {
      address = arg;
    } else if (model == "a") {
      model = arg;
    } else {
      PrintUsage(argv[0]);
      return 1;
    }
  }

  if (address.empty()) {
    PrintUsage(argv[0]);
    return 1;
  }

  model = ToLower(model);

  if (model == "a") {
    return Run<y1_model::A>(address);
  }
  if (model == "m") {
    return Run<y1_model::M>(address);
  }

  std::cerr << "Unknown model: " << model << std::endl;
  PrintUsage(argv[0]);
  return 1;
}
